filebrowser: 2.23.0 -> 2.28.0, add service module

christoph-heiss opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Updates the `filebrowser` package from 2.23.0 to 2.28.0 and adds an appropriate service module for it.

Changelog: https://github.com/filebrowser/filebrowser/compare/v2.23.0...v2.28.0

cc @nielsegberts As this package has not been updated in nearly a year, are you still interested?

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

christoph-heiss committed:
filebrowser: 2.23.0 -> 2.28.0

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
filebrowser: add myself as maintainer

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
nixos/filebrowser: init module

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
nixos/tests/filebrowser: init test module

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

fin444 commented:
I haven't done rigorous testing, and I'm not experienced enough to provide full review of a module, but it is working well for me and seems good overall.

Filebrowser's configuration has setting the host/port and socket under the same field, but that feels a bit weird IMO. Additionally, sockets typically (AFAIK) get put in `/run/service/service.sock` (`/run/filebrowser/filebrowser.sock`), however there are no permissions to this location unless

```
systemd.services.filebrowser.serviceConfig.RuntimeDirectory = "filebrowser";
```

Though, this directory would be useless if sockets are turned off. Maybe `services.filebrowser.useSocket` or similar, and then set host/port otherwise?

Also, I think `services.filebrowser.disableCommandRunner` would be more clear than `services.filebrowser.disableExec`, even if it doesn't match the CLI args.

christoph-heiss commented:
> I haven't done rigorous testing, and I'm not experienced enough to provide full review of a module, but it is working well for me and seems good overall.

Thanks for the review! I'll happily discuss how to make this module better.

> 
> Filebrowser's configuration has setting the host/port and socket under the same field, but that feels a bit weird IMO. 

Well, while writing this I pondered over this quite a bit myself, since filebrowser itself takes a separate option for the socket. But most services take either a IP-address or a socket path for the address, since the type can be determined quite reliably anyway. 

Additionally, sockets typically (AFAIK) get put in `/run/service/service.sock` (`/run/filebrowser/filebrowser.sock`), however there are no permissions to this location unless

> 
> ```
> systemd.services.filebrowser.serviceConfig.RuntimeDirectory = "filebrowser";
> ```
> 
> Though, this directory would be useless if sockets are turned off. Maybe 
`services.filebrowser.useSocket` or similar, and then set host/port otherwise?

Most modules AFAIK just put it under `/run` directly anyway - and since we'll only have one file, I would avoid the extra directory too.
We could also do it how e.g. [`services.postgrey`](https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/services/mail/postgrey.nix#L64) does it, although I'd name the option `address`. 
Such that one could do 
```nix
services.filebrowser.listenAddress = { address = "0.0.0.0"; port = 8080; }; # or
services.filebrowser.listenAddress = { socket = "/run/filebrowser.sock"; mode = "0666"; };
```

> 
> Also, I think `services.filebrowser.disableCommandRunner` would be more clear than `services.filebrowser.disableExec`, even if it doesn't match the CLI args.

Seems sensible. :+1: Personally I don't have a strong opionon on this tho.

fin444 commented:
That sounds like a good method, it still feels a bit weird but I'm not sure there's a better way. For the socket path, `/run/filebrowser.sock` would also work, I just don't know about the permissions. I tried setting the socket to that, but it failed off permissions, which is why I thought it needed its own directory created by systemd. Reading through Postgrey's module, I don't really see how they solve this issue but clearly they do.

christoph-heiss committed:
filebrowser: drop nielsegberts from maintainers

This package hasn't been updated in ~1.5 years and the maintainer does
not seem to respond.

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss commented:
@fin444 Sorry for the long feedback loop!

In any case, I have now simply removed the socket option, since it's quite a bit of work and needs a more thought-out plan as can be seen. I'd rather not get this PR stuck on this forever.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3816

AndersonTorres commented:
Marking @nielsegberts here https://github.com/NixOS/nixpkgs/issues/290642

AndersonTorres commented on nixos/modules/services/web-apps/filebrowser.nix:
```diff
@@ -0,0 +1,179 @@
+          ${pkgs.coreutils}/bin/mkdir -p ${toString cfg.rootDir}
+        '';
+
+        ExecStart = "${pkgs.filebrowser}/bin/filebrowser --config ${configFile}";
```
Parameterize the package, using `cfg.package`.

SuperSandro2000 commented on nixos/modules/services/web-apps/filebrowser.nix:
```diff
@@ -0,0 +1,181 @@
+          ${pkgs.coreutils}/bin/mkdir -p ${toString cfg.rootDir}
+        '';
+
+        ExecStart = "${cfg.package}/bin/filebrowser --config ${configFile}";
```
```suggestion
        ExecStart = "${lib.getExe cfg.package} --config ${configFile}";
```

SuperSandro2000 commented on nixos/modules/services/web-apps/filebrowser.nix:
```diff
@@ -0,0 +1,181 @@
+        DynamicUser = enableDynamicUser;
+
+        # Basic hardening
+        NoNewPrivileges = "yes";
+        PrivateTmp = "yes";
+        PrivateDevices = "yes";
+        DevicePolicy = "closed";
+        ProtectSystem = "strict";
+        ProtectHome = "tmpfs";
+        ProtectControlGroups = "yes";
+        ProtectKernelModules = "yes";
+        ProtectKernelTunables = "yes";
+        RestrictAddressFamilies = "AF_UNIX AF_INET AF_INET6";
+        RestrictNamespaces = "yes";
+        RestrictRealtime = "yes";
+        RestrictSUIDSGID = "yes";
+        MemoryDenyWriteExecute = "yes";
+        LockPersonality = "yes";
```
In nixpkgs we mostly use true instead of "yes"

SuperSandro2000 commented on nixos/modules/services/web-apps/filebrowser.nix:
```diff
@@ -0,0 +1,181 @@
+      type = types.str;
+      default = "filebrowser";
+      description = ''
+        If `services.filebrowser.user` is set, this must be set as well. Must
```
Can't we create the default user?

SuperSandro2000 commented on nixos/modules/services/web-apps/filebrowser.nix:
```diff
@@ -0,0 +1,181 @@
+      description = "Number of image processes.";
+    };
+
+    disableThumbnails = mkEnableOption "Disable image thumbnails.";
+    disablePreviewResize = mkEnableOption "Disable resize of image previews.";
+    disableCommandRunner = mkEnableOption "Disable the Command Runner feature.";
+    disableTypeDetectionByHeader = mkEnableOption
+      "Disable file type detection by reading file headers.";
```
The description of those will be prefixed with ``Wether to enable`` which I don't think the description considered.

SuperSandro2000 commented on nixos/modules/services/web-apps/filebrowser.nix:
```diff
@@ -0,0 +1,181 @@
+
+    baseUrl = mkOption {
+      type = types.nullOr types.str;
+      default = null;
```
Shouldn't this default to / ?

SuperSandro2000 commented on nixos/modules/services/web-apps/filebrowser.nix:
```diff
@@ -0,0 +1,181 @@
+      cert = cfg.tlsCertificate;
+      key = cfg.tlsCertificateKey;
+    })
+    // (optionalAttrs (cfg.baseUrl != null) { baseurl = cfg.baseUrl; })
```
optionalAttrs doesn't require sourounding brackets

SuperSandro2000 commented on nixos/modules/services/web-apps/filebrowser.nix:
```diff
@@ -0,0 +1,181 @@
+      key = cfg.tlsCertificateKey;
+    })
+    // (optionalAttrs (cfg.baseUrl != null) { baseurl = cfg.baseUrl; })
+    // (optionalAttrs (cfg.cacheDir != null) { cache-dir = cfg.cacheDir; })
+    // (optionalAttrs cfg.disableThumbnails { disable-thumbnails = true; })
+    // (optionalAttrs cfg.disablePreviewResize { disable-preview-resize = true; })
+    // (optionalAttrs cfg.disableCommandRunner { disable-exec = true; })
+    // (optionalAttrs cfg.disableTypeDetectionByHeader { disable-type-detection-by-header = true; })
```
Why not use an rfc 42 freeform setting type for this?

SuperSandro2000 commented on nixos/modules/services/web-apps/filebrowser.nix:
```diff
@@ -0,0 +1,181 @@
+    cfg.tlsCertificate != null;
+
+  configFile = pkgs.writeText "filebrowser-config.json" (generators.toJSON {} ({
+      database = "/var/lib/filebrowser/filebrowser.db";
```
We use 2 spaces to intend nix code

```suggestion
    database = "/var/lib/filebrowser/filebrowser.db";
```

SuperSandro2000 commented on nixos/modules/services/web-apps/filebrowser.nix:
```diff
@@ -0,0 +1,181 @@
+    assert cfg.user != "filebrowser" -> cfg.group != "filebrowser";
+    cfg.user == "filebrowser";
+  enableTls =
+    assert cfg.tlsCertificate != null -> cfg.tlsCertificateKey != null;
```
This should be a module level assertion

SuperSandro2000 commented on nixos/modules/services/web-apps/filebrowser.nix:
```diff
@@ -0,0 +1,181 @@
+                mkOption mkPackageOption optional optionalAttrs;
+
+  cfg = config.services.filebrowser;
+  enableDynamicUser =
+    assert cfg.user != "filebrowser" -> cfg.group != "filebrowser";
+    cfg.user == "filebrowser";
```
I find it kinda strange and unexpected that DynamicUser gets disabled when the user and group settings are changed.

AndersonTorres commented:
PING
